scylla-cdc-base/src/main/java/com/scylladb/cdc/cql/driver3/Driver3MasterCQL.java

```diff
@@ -20,40 +23,176 @@
 import com.datastax.driver.core.ResultSetFuture;
 import com.datastax.driver.core.Row;
 import com.datastax.driver.core.Session;
+import com.datastax.driver.core.Statement;
 import com.google.common.base.Preconditions;
+import com.google.common.flogger.FluentLogger;
 import com.google.common.util.concurrent.FutureCallback;
 import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
 import com.scylladb.cdc.cql.BaseMasterCQL;
+import com.scylladb.cdc.model.FutureUtils;
 
 public final class Driver3MasterCQL extends BaseMasterCQL {
+    private static final FluentLogger logger = FluentLogger.forEnclosingClass();
 
     private final Session session;
-    private final PreparedStatement fetchSmallestGenerationAfterStmt;
-    private final PreparedStatement fetchStreamsStmt;
+
+    // (Streams description table V2)
+    //
+    // PreparedStatements for querying in clusters with
+    // system_distributed.cdc_generation_timestamps
+    // and system_distributed.cdc_streams_descriptions_v2 tables.
+    private PreparedStatement fetchSmallestGenerationAfterStmt;
+    private PreparedStatement fetchStreamsStmt;
+    private boolean foundRewritten = false;
+
+    // (Streams description table V1)
+    //
+    // PreparedStatements for querying in clusters WITHOUT
+    // system_distributed.cdc_generation_timestamps
+    // and system_distributed.cdc_streams_descriptions_v2 tables.
+    private PreparedStatement legacyFetchSmallestGenerationAfterStmt;
+    private PreparedStatement legacyFetchStreamsStmt;
 
     public Driver3MasterCQL(Session session) {
         Preconditions.checkNotNull(session);
         this.session = session;
-        fetchSmallestGenerationAfterStmt = session.prepare(getFetchSmallestGenerationAfter());
-        fetchStreamsStmt = session.prepare(getFetchStreams());
```

Comment:
How does concurrency work with this being lazily initialised? Is it guaranteed that we won't prepare the same statement twice?

Reply:
What I mean is that fetchSmallestGenerationAfterStmt and fetchStreamsStmt were final previously and now they are not final any more so it's a shared mutable state.

Reply:
The current users of MasterCQL (Master) effectively use MasterCQL sequentially (.get() on futures, never starting many futures at once), so I think there are no cases where the same statement is prepared by two futures simultaneously.

Reply:
This seems fragile. Please at least add a comment explaining this.

Reply:
does it even matter if the same statement is prepared twice? I think not...

Reply:
> does it even matter if the same statement is prepared twice? I think not...

That was also my thinking.